### ISSUE TITLE
working with flags; adding, modifying, etc

### ISSUE THREAD

**Issue by richardsc** [dankelley/oce#1408]
The addition of the `@metadata$flags` field, along with some generic `handleFlags()` functions, has (I think) greatly increased the utility of oce for working with data. I think we're at a point now where we should be thinking about what the next steps are.

While the `flags` field is created for all new oce objects, it is only populated for those files which we know to recognize the existence of flags in the raw data (`argo`, `ctd`/`odf`, `section` ...). However, it makes sense to have the facility to be able to **add and modify** flags within oce objects read from raw files, along the way to producing a processed data set.

The present use case for this involves work that @Echisholm21 is doing with ADCP data (which currently has *no* flag functionality), but I think that with little work this could be made general for all oce objects.

I see an example use case looking something like (not real code, just making this up):
```r
d <- read.oce('rawDataFile.dat') #read a raw file with no flag information
d <- addFlags(d, value=1) #populate the flag field with a default value, e.g. for "good" data
d <- setFlags(d, field='salinity', value=4, condition={d[['salinity']] < 15}) #use a logical to identify values to change
dgood <- handleFlags(d)
```

Discuss.

**Comment by dankelley** [comment]
A f2f may be the best way to figure out the syntax for setting flags.  It will likely be done in a new generic, and that means that syntax choices will ripple through the system, so we will need to fill up a blackboard with some "what if" scenarios.

Whatever is done will have to be sensible across a range of object types.  For example, it would be nice if the syntax would provide a comfortable tool for the task of altering pre-existing flags.

**Comment by dankelley** [comment]
Q for @richardsc have you ever seen character-based flags? At the moment, oce is set up with numeric values in mind. But, if some institutes use characters, I might need to go into the code to see whether we need some tricks. This came up because I was thinking that maybe flags should be `raw`, so they only use 1 byte each. Using `raw` could (somewhat) alleviate my concern about flag bloat for adp objects, but I still argue that by default they should be empty. (After all, oce tries to be like the data. Hydrographic data as supplied on international archive sites commonly have flags. But no adp that I know of has flags, in data supplied by the manufacturer.)

I fear these GH comments are already getting a bit confusing. I will likely code a trial `handleFlags` on the weekend and I propose that it get played with a bit, to see what "feels" right. If we are saying that "bad" is "bad", then the default action should be to set velo to NA.  And, I just checked, and we have *no* choice in the matter, viz
```R
> a<-as.raw(c(0x01, 0x02))
> a
[1] 01 02
> a[1]<-NA
Error in a[1] <- NA : 
  incompatible types (from logical to raw) in subassignment type fix
```
so we can't set `a`, `g` and `q` to NA, anyhow.

**Comment by dankelley** [comment]
I like the discussion of functions to set and modify flags, but it might be worth pointing out that we can do that using the `[[` mechanism, as below. (This works in the "develop" branch, commit c574b7f3fe033797919afae6bc3a15155c8bb985 of a few minutes ago.)

What we could have with new functions is more granularity. The `[[` mechanism only works to update the *whole* entity in question (e.g. below I create a whole array of flags, named `vFlag` and then I shove this array nto the object by using `adp[["vFlag"]] <- vFlag`).  This is because e.g. `adp[["vFlag"]][1,1,1]<-1` will not work, because the `[[` operator does not yield what's called an "l-value" in computing parlance.  And letting it take e.g. `adp[["vFlag", 1, 1, 1]] <- 1` is not really a choice (I think) because `[[` is a generic and it has to obey the base R meaning of `[[` in terms of argument list (called "signature").

Sorry this is too long -- I find GH comments a bit hopeless -- but the question for EC and CR is whether the syntax below looks reasonable.  If so, I'll put this example in the vignette.

NOTE: in the example, I made it very sensitive to `v[,,4]` values, just so the eye could see the difference in the lower panels of the plot. This is because the instrument had already set a lot of the BIO-flagged data as bad, during its internal processing. (Clark may know how that processing worked -- I wasn't part of setting up the instrument.)

```R
library(oce)
data(adp)
v <- adp[["v"]]
g <- adp[["g", "numeric"]]
dim <- dim(v)
vFlag <- array(0, dim=dim)

gThreshold <- 25
v4Threshold <- 0.05 # BIO suggests 0.45 but make it dramatic for a test
for (i in 1:3)
    vFlag[,,i] <- ifelse((g[,,i]+g[,,4])<gThreshold | v[,,4]>v4Threshold, 1, 0)

adp[["vFlag"]] <- vFlag

par(mfcol=c(2, 3))
for (i in 1:3) {
    plot(adp, which=i)
    plot(handleFlags(adp), which=i)
}
```

**Comment by richardsc** [comment]
Sorry for not responding to this earlier -- @EChisholm21 and I had a good meeting today and discussed various aspects of what would be a good scheme, and what might be specific to ADCP processing. I think for oce we want something that is quite general, and can be applied to different objects with different data fields and dimensions, and if something more refined is required (building on the oce paradigm) that can be done separately.

Definitely a f2f next week would be good to sort through this. But there is no rush, better to do it right.

A few notes from our meeting:

1. I think that fields like percent good, correlation, and backscatter will basically never need flags. This is because they are recorded by the instrument as a way of *doing* quality control, so it makes no sense to apply QC flags to them. Ultimately, for an `adp` object, the flags should be applied to the `v` array (and maybe `pressure`, `temperature`, etc but that's different from this discussion).

2. Maybe I forgot, maybe I never knew, or maybe you *just* added the
    ```r
    adp[['vFlag']] <- vFlag
    ```
    functionality. I agree that is possibly cleaner than coding up some other generic function, though it does require making sure that the flag matrix is dimensionally correct before inserting it as a "flag" field. It's also not as intuitive as a function, since it's harder to add readable information to a help file when it's buried in the man page for `[[,adp-method`.

    The idea of having a separate function to add granularity is worth discussing -- e.g. the `adp[['vFlag']]` approach requires updating the entire flag array as you pointed out, rather than just values that meet a certain criterion/criteria.

3. I don't know of any files that support character flags. @Echisholm21 and I talked about some `adp` specific cases that could benefit from a finer level of granularity in the flagging scheme, but my recommendation was to start with something consistent with current flag practices (numeric, single value) and we could reconsider once that is mastered.

    For example, one might want to know from the flags on velocity *which* field/test was used to flag the data as bad, separating out the percentgood/correlation/backscatter/error velocity failures.

4. If you find the GH issues comments confusing, it's only because you can't help but write several exhaustive treaties on the subject withing the span of a few hours, and then can't find the original thought that started it all by the end ... :smile: 

Have a great weekend!

**Comment by dankelley** [comment]
I've been working away on this (as discussed f2f yesterday) in the "dk" branch.  It will take a few days more to see how things "feel" in practice, and as part of that I think I'll make a vignette. The vignette can address things that span several functions, which will help because there will be interlocking functions, e.g. if `setFlagScheme()` has been used on an object, then `setFlags()` can refer to flags by *either* numerical value or by name, **but** if no scheme has been set up, then flags can only be referred to by numerical value. A vignette can help a lot with such things because it can show different working schemes.  I may not push to "develop" on the upcoming Saturday, which would have been my plan normally, because a lot of this is tricky (e.g. I fell into a "recursive default argument reference" rabbit hole this morning).

**Comment by dankelley** [comment]
Closing, since this has been addressed in the "develop" branch, commit 3ccd35cc70479d69ec78abb0570342a7715346d7, which was merged from the "dk" branch earlier today.
